modules/renter/proto/merkleroots.go

```diff
@@ -0,0 +1,368 @@
+	// its elements into mr.uncachedRoots.
+	mr.cachedSubTrees = mr.cachedSubTrees[:len(mr.cachedSubTrees)-1]
+	rootIndex := len(mr.cachedSubTrees) * merkleRootsPerCache
+	roots, err := mr.merkleRootsFromIndex(rootIndex, mr.numMerkleRoots)
+	if err != nil {
+		return errors.AddContext(err, "failed to read cached tree's roots")
+	}
+	mr.uncachedRoots = append(mr.uncachedRoots, roots...)
+	return nil
+}
+
+// insert inserts a root by replacing a root at an existing index.
+func (mr *merkleRoots) insert(index int, root crypto.Hash) error {
+	if index > mr.numMerkleRoots {
+		return errors.New("can't insert at a index greater than the number of roots")
+	}
+	if index == mr.numMerkleRoots {
+		return mr.push(root)
+	}
+	// Replaced the root on disk.
+	_, err := mr.file.WriteAt(root[:], fileOffsetFromRootIndex(index))
+	if err != nil {
+		return errors.AddContext(err, "failed to insert root on disk")
+	}
+
+	// Find out if the root is in mr.cachedSubTree or mr.uncachedRoots.
+	i, cached := mr.isIndexCached(index)
+	// If the root was not cached we can simply replace it in mr.uncachedRoots.
+	if !cached {
+		mr.uncachedRoots[i] = root
+		return nil
+	}
+	// If the root was cached we need to rebuild the cache.
+	if err := mr.rebuildCachedTree(i); err != nil {
+		return errors.AddContext(err, "failed to rebuild cache for inserted root")
+	}
+	return nil
+}
+
+// isIndexCached determines if the root at index i is already cached in
+// mr.cachedSubTree or if it is still in mr.uncachedRoots. It will return true
+// or false and the index of the root in the corresponding data structure.
+func (mr *merkleRoots) isIndexCached(i int) (int, bool) {
+	if i/merkleRootsPerCache == len(mr.cachedSubTrees) {
+		// Root is not cached. Return the false and the position in
+		// mr.uncachedRoots
+		return i - len(mr.cachedSubTrees)*merkleRootsPerCache, false
+	}
+	return i / merkleRootsPerCache, true
+}
+
+// lenFromFile returns the number of merkle roots by computing it from the
+// filesize.
+func (mr *merkleRoots) lenFromFile() (int, error) {
+	stat, err := mr.file.Stat()
+	if err != nil {
+		return 0, err
+	}
+	size := stat.Size()
+	// If we haven't written a single root yet we just return 0.
+	if size < contractHeaderSize {
+		return 0, nil
+	}
+
+	// Sanity check contract file length.
+	if (size-contractHeaderSize)%crypto.HashSize != 0 {
+		return 0, errors.New("contract file has unexpected length and might be corrupted")
+	}
+	return int((size - contractHeaderSize) / crypto.HashSize), nil
+}
+
+// len returns the number of merkle roots. It should always return the same
+// number as lenFromFile.
+func (mr *merkleRoots) len() int {
+	return mr.numMerkleRoots
+}
+
+// appendRoot appends a root to the in-memory structure of the merkleRoots. If
+// the length of the uncachedRoots grows too large they will be compressed into
+// a cachedSubTree.
+func (mr *merkleRoots) appendRoot(root crypto.Hash) {
```

Comment:
may want to call this `appendRootMemory` or something, so that the function name clearly indicates that it only modifies the in-memory structure.